[SideNav] Integrate existing navigation tree

Dosant opened the pull request:
## Summary


> [!IMPORTANT]
> A lot of lines added because of a testing tree: `src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/tests/mock_security_tree.json` . The actual change is actually quite small—just 500 lines!


> [!TIP]
> A-la-carte deployement, make sure to use one of spaces with solution navigation: https://dosant-pr-230102-d-2025-07-31-integrate-ia.kbndev.co/s/space/  | elastic / changeme


This PR integrates the new side nav with the existing navigation tree structure by mapping the hierarchical navigation definition to the new format. 

To see the new side nav, enable feature flag and use solution navigation in a space or serverless:

```
feature_flags.overrides:
  core.chrome.projectSideNav: 'v2'
```


**This is a good start. After this, we will polish each tree individually and add the missing side navigation features. It's important to note that there is no active node handling yet! So the sidenav's active state can get out of sync with the actual active path.**


### **Structural Assumptions and Mapping**
- **Root node (1st level)** is used for the "logo" item and application branding
- **2nd level nodes** are transformed into primary navigation items:
  - Accordion nodes are flattened (not supported) - their children become primary items
  - Nodes without links that aren't panel openers are treated as section dividers - their children are flattened
  - panelOpener nodes create flyout secondary navigation panels
- **3rd level** is used for secondary navigation (children of panelOpener):
  - If all 3rd level items have links, they're treated as menu items and wrapped in a single section
  - If some don't have links, they're treated as section headers with their children becoming menu items
- **Footer** is limited to 5 items maximum (extras are dropped with warning)


### **Special Handling and Fallbacks**
- Missing properties get sensible fallbacks with warnings (titles, hrefs, icons)
  - The fallback icons are incorrect; it's just a quick mapping that needs to be replaced.
- Primary items missing hrefs try to use the first child item's href as fallback (tbd)
- Test subjects are preserved for automated testing
 
### **Unsupported Features**

The transformation warns when encountering unsupported features:

- Accordion items (flattened instead)
- Badge options
- spaceBefore property
- Custom render functions
- "Open in new tab" settings (external links still work)
- Recently accessed nodes

### **Warning System**

A batched warning system collects and logs all navigation structure issues at once rather than flooding the console, making debugging easier for developers integrating with the new navigation.

<img width="1011" height="670" alt="Screenshot 2025-08-01 at 14 30 33" src="https://github.com/user-attachments/assets/78aaa0f3-495d-4599-8797-0ef2f2b8264d" />



### `both` Feature Flag 😆

For easier testing and comparison, I've added a `both` value for the side nav feature flag. This lets you see both side navs at the same time.

```
feature_flags.overrides:
  core.chrome.projectSideNav: 'both'
```

<img width="1507" height="865" alt="Screenshot 2025-08-01 at 14 32 25" src="https://github.com/user-attachments/assets/171f4b73-30a3-49ec-860d-bb6658b233ae" />

Dosant committed:
basic IA integration

Dosant committed:
add a comment

Dosant committed:
fixes

Dosant committed:
don’t pass links to items with children

Dosant committed:
improvements

Dosant committed:
Merge branch 'main' into d/2025-07-31-integrate-ia

tsullivan commented on src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/navigation.tsx:
```diff
@@ -0,0 +1,405 @@
+        return null;
+      }
+
+      const items = filterEmpty(navNode.children.flatMap(toMenuItem));
```
Nit, for consideration: should we convert accordion nodes to `panelOpener` instead? Either way is a bit hacky. Maybe your method actually is better as it would incentivize more urgent clean-up.

Dosant replied:
That discussion is resolved, but it still seems that flatten is closer to the end result.

tsullivan commented on src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/navigation.tsx:
```diff
@@ -0,0 +1,405 @@
+  return { logoItem, navItems: { primaryItems, footerItems } };
+};
+
+// =====================
+// Utilities & Helpers
+// =====================
```
Nit: separate file for these? I believe it could help to keep this file smaller, even if only slightly

tsullivan commented on src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/navigation.tsx:
```diff
@@ -0,0 +1,405 @@
+    label: warnIfMissing(logoNode, 'title', 'Kibana'),
+  };
+
+  const toMenuItem = (navNode: ChromeProjectNavigationNode): MenuItem[] | MenuItem | null => {
```
`toMenuItem` can be a top-level function; it does not need to be enclosed

Dosant replied:
Also, a preference. `toMenuItem` is just an internal implementation detail of that mapper, and we don't really need to keep it outside.

tsullivan commented on src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/redirect_app_links.tsx:
```diff
@@ -0,0 +1,36 @@
+      navigateToUrl={application.navigateToUrl}
+      currentAppId={currentAppId}
+      // reset default redirect app links styles
+      css={{ display: 'flex', height: '100%', flexDirection: 'row' }}
```
Nit: Maybe update `src/platform/packages/shared/shared-ux/link/redirect_app/impl/src/redirect_app_links.component.tsx` to use a merge or make it more obvious that the `css` is getting overwritten from higher up?
```tsx
// ... /link/redirect_app/impl/src/redirect_app_links.component.tsx
export const RedirectAppLinks: FC<RedirectAppLinksComponentProps> = ({
  children,
  navigateToUrl,
  currentAppId,
  css: customStyles,
  ...containerProps
}) => {
  // ...

  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events
    <div
      onClick={handleClick}
      ref={containerRef}
      css={[redirectAppLinksStyles, customStyles]}
      data-test-subj="kbnRedirectAppLink"
      {...containerProps}
    >
      {children}
    </div>
  );
};

```

tsullivan reviewed (commented):
Left a few comments. Let me know what you think and I can approve.

This PR is full of great stuff! I like that everything being done to smooth over the existing structures is logged as a warning so that teams can clearly see and tackle the issues on their own. The cleanup will bring in changes to the nav structures that will create nice-looking results in both V1 and V2 of the side nav.

Dosant committed:
Clean up

Dosant commented:
@tsullivan, thanks for the review! It's very helpful. Regarding the code structure comments, as I mentioned, I prefer to keep everything in one place. However, the main change I made was splitting the component/hook and state mapping into separate files, which I think makes it a lot better. 👍

Dosant committed:
improve

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/326374)
* Commit: edb1759b9cdd34c1ec3f1465246da50d367764de

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `core` | 524 | 528 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `core` | 85.3KB | 110.7KB | +25.4KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `core` | 506.9KB | 482.8KB | -24.1KB |

<details><summary>Unknown metric groups</summary>

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `core` | 2 | 3 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `@kbn/core-chrome-browser-internal` | 6 | 7 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/4ef6c5d08050b7e689b87d3dfd171941c5665cc5) | [after](https://github.com/elastic/kibana/commit/edb1759b9cdd34c1ec3f1465246da50d367764de) | diff |
| --- | --- | --- | --- |
| `@kbn/core-chrome-browser-internal` | 6 | 7 | +1 |


</details>


### History
* :green_heart: [Build #325893](https://buildkite.com/elastic/kibana-pull-request/builds/325893) succeeded 5028c03d34b354fe2caa9f9f5354e6fdc9d4a3cf
* :broken_heart: [Build #325531](https://buildkite.com/elastic/kibana-pull-request/builds/325531) failed 5d3257fd6588244d90468e5d49c8af0b8da03e3b
* :green_heart: [Build #325252](https://buildkite.com/elastic/kibana-pull-request/builds/325252) succeeded f325d000f822066d6b81dccd5fdf011078815c87

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326374","number":326374,"commit":"edb1759b9cdd34c1ec3f1465246da50d367764de"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325893","number":325893,"commit":"5028c03d34b354fe2caa9f9f5354e6fdc9d4a3cf"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325531","number":325531,"commit":"5d3257fd6588244d90468e5d49c8af0b8da03e3b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325252","number":325252,"commit":"f325d000f822066d6b81dccd5fdf011078815c87"}],"number":326374}
buildkite-pr-comment-->

Dosant committed:
Primary menu items must always be links

Primary menu items must always be links.
If they open a panel, they should use the link from the first child.
If a solution provides its own link and children, we will warn that providing a link is incorrect and will ignore it, falling back to the link from the first child.

tsullivan reviewed (approved):
LGTM. Thanks for responding to my feedback!
